src/client.ts

```diff
@@ -1,149 +1,224 @@
+  }
 
-  public async verifyAuthentication(): Promise<void> {
-    // TODO make the most light-weight request possible to validate
-    // authentication works with the provided credentials, throw an err if
-    // authentication fails
-    const request = new Promise((resolve, reject) => {
-      http.get(
-        {
-          hostname: 'localhost',
-          port: 443,
-          path: '/api/v1/some/endpoint?limit=1',
-          agent: false,
-          timeout: 10,
-        },
-        (res) => {
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
+  private withBaseUri(path: string): string {
+    return `https://api.malwarebytes.com/${path}`;
+  }
+
+  private async request(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    body?: any,
+  ): Promise<Response> {
+    if (!APIClient.accessToken) {
+      await this.getAuthenticationToken();
+    }
+
+    return await fetch(uri, {
+      method,
+      headers: {
+        Authorization: `Bearer ${APIClient.accessToken}`,
+        AccountId: this.accountId,
+        'Content-Type': 'application/json',
+      },
+      body: body || null,
+    });
+  }
+
+  private async makePaginatedRequest<K extends string, V>(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    resource: K,
+    body?: any,
+  ): Promise<V[]> {
+    const results: V[] = [];
+    let nextCursor = '';
+
+    do {
+      const response = await this.request(
+        uri,
+        method,
+        JSON.stringify({
+          ...body,
+          page_size: 10,
+          next_cursor: nextCursor,
+        }),
       );
+
+      const responseJson: PaginatedDataResponse<K, V[]> = await response.json();
+
+      if (responseJson) {
+        nextCursor = responseJson.next_cursor;
+        results.push(...responseJson[resource]);
+      }
+    } while (nextCursor && nextCursor.length > 0);
+
+    return results;
+  }
+
+  private async getAuthenticationToken() {
+    const getAuthTokenEndpoint = this.withBaseUri('oauth2/token');
+    const response = await fetch(getAuthTokenEndpoint, {
+      method: 'POST',
+      headers: {
+        Authorization: `Basic ${Buffer.from(
+          `${this.clientId}:${this.clientSecret}`,
+        ).toString('base64')}`,
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: qs.stringify({
+        scope: 'read',
+        grant_type: 'client_credentials',
+      }),
     });
 
     try {
-      await request;
-    } catch (err) {
+      const body = await response.json();
+      APIClient.accessToken = body.access_token;
+    } catch (e) {
       throw new IntegrationProviderAuthenticationError({
-        cause: err,
-        endpoint: 'https://localhost/api/v1/some/endpoint?limit=1',
-        status: err.status,
-        statusText: err.statusText,
+        cause: new Error('Provider authentication failed'),
```

Comment:
The `cause` property is optional on this Error, and is used in order to pass any thrown error, if applicable. In this case, you should pass in `e`:

```suggestion
        cause: e,
```